File: bindings/python/orchustr/forge.py
from __future__ import annotations

import asyncio


class ForgeRegistry:
    def __init__(self) -> None:
        self._tools: dict[str, object] = {}

    def register(self, name: str, handler) -> None:
        self._tools[name] = handler

    async def import_from_mcp(self, client) -> int:
        for tool in await client.list_tools():
            async def proxy(args, tool_name=tool["name"]):
                return await client.invoke_tool(tool_name, args)
            self._tools[tool["name"]] = proxy
        return len(self._tools)

    async def invoke(self, name: str, args: dict) -> object:
        if name not in self._tools:
            raise KeyError(f"unknown tool: {name}")
        result = self._tools[name](args)
        if asyncio.iscoroutine(result):
            return await result
        return result

File: bindings/dart/lib/src/forge.dart
import "dart:async";

import "mcp.dart";
import "types.dart";

typedef ForgeToolHandler = FutureOr<Object?> Function(JsonObject args);

final class ForgeRegistry {
  final Map<String, ForgeToolHandler> _tools = <String, ForgeToolHandler>{};

  void register(String name, ForgeToolHandler handler) {
    _tools[name] = handler;
  }

  Future<int> importFromMcp(NexusClient client) async {
    for (final tool in await client.listTools()) {
      final toolName = "${tool["name"]}";
      _tools[toolName] = (JsonObject args) => client.invokeTool(toolName, args);
    }
    return _tools.length;
  }

  Future<Object?> invoke(String name, JsonObject args) async {
    final handler = _tools[name];
    if (handler == null) {
      throw StateError("unknown tool: $name");
    }
    return await Future<Object?>.value(handler(args));
  }
}
